flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/FileBasedBufferIterator.java

```diff
@@ -85,6 +90,6 @@ private int read(byte[] buffer) {
 
 	@Override
 	public void close() throws Exception {
-		closeAll(stream, file::release);
+		closeAll(stream, file::release, () -> buffersToClose.forEach(Buffer::recycleBuffer));
```

Comment:
Why are we recycling buffers at the end, on close and not on the fly? It can mean GBs of extra (unnecessary) memory usage if we keep all unconsumed buffers from different channels. The ownership of the buffers returned from the iterator should belong to the caller of `next()` method.

Reply:
I've changed it to recycle only the last buffer in `close`.
I wanted to avoid extra complexity but it turned out not that complicated.

(the reason why buffers need to be recycled by iterator rather than by client is that for in-memory collections we always want to recycle all the elements, even if iteration wasn't started)

Reply:
I don't understand it how is it working right now. Where are the remaining (non last) buffers recycled? And why do you have to handle the last buffer differently? As I mentioned before, I think:
> The ownership of the buffers returned from the iterator should belong to the caller of next() method.

so buffers' ownership should be handed over, and owner should be always responsible to recycle it when it's done using it.

Reply:
> I don't understand it how is it working right now. Where are the non last buffers recycled? 

In the beginnning of `FileBasedBufferIterator.next()`.

> And why do you have to handle the last buffer differently?

Because `next()` won't be called anymore.

> The ownership of the buffers returned from the iterator should belong to the caller of next() method.

The client doesn't know whether the iterator is lazy or not. So if failure happens before iteration, it would have to read the whole file just to recycle the buffers.

> I guess we are avoiding memory leaks right now, only because the returned Buffers are not pooled and using simple FreeingBufferRecycler?

No, the buffers are recycled explicitly (see above).

Reply:
> The client doesn't know whether the iterator is lazy or not. So if failure happens before iteration, it would have to read the whole file just to recycle the buffers.

? Why? As I wrote before, the buffers that were not yet returned, should be recycled by it's current owner - in that case `CloseableIterator`. If the buffers were even not yet created, because it's a lazy iterator, even better - nothing to do in the `close()`. 

> In the beginnning of FileBasedBufferIterator.next().

If buffers were returned from the iterator, iterator is no long it's owner and they shouldn't be recycled by the iterator.

Reply:
I think you are right, these ownership rules are more clear (though implementation will be a bit more complex and less optimal in collections iterators).
I'll update the PR.

Reply:
Done.